src/preferences.c

```diff
@@ -210,319 +67,354 @@ int preferences_guess_best_font_size(int target_cols){
-	if(!preferences){
-		preferences_init();
-	}
-	int value;
-  config_lookup_int(preferences, pref, &value);
-  return value;
+	/* call the rendering function later after SDL init*/
+	menu->surface = NULL;
+
+	return menu;
 }
 
-int preferences_get_bool(char* pref){
-	if(!preferences){
-		preferences_init();
-	}
-	int value;
-	config_lookup_bool(preferences, pref, &value);
-	return value;
+void destroy_preferences(pref_t *pref) {
+	free(pref->font_path);
+	free(pref->tty_encoding);
+
+	free(pref->text_color);
+	free(pref->background_color);
+	free(pref->metamode_hitbox);
+	
+	keymap_t *m = pref->metamode_keys;
+	while (m->to != NULL) { free(m->to); ++m; }
+	free(pref->metamode_keys);
+	
+	m = pref->metamode_sticky_keys;
+	while (m->to != NULL) { free(m->to); ++m; }
+	free(pref->metamode_sticky_keys);
+	
+	m = pref->metamode_func_keys;
+	while (m->to != NULL) { free(m->to); ++m; }
+	free(pref->metamode_func_keys);
+
+	destroy_symmenu(pref->main_symmenu);
+	
+	free(pref->keyhold_actions_exempt);
+
+	free(pref);
 }
 
-int preferences_get_int_array(char* pref, int* fillme, int length){
-	config_setting_t *root, *setting;
-	if(!preferences){
-		preferences_init();
+#define DEFAULT_LOOKUP(type, conf, path, target, defval)	  \
+	do { if (CONFIG_TRUE != config_lookup_##type(conf, path, &(target))) { target = defval; } } while(0)
+pref_t *read_preferences(const char* filename) {
+	pref_t *prefs = calloc(1, sizeof(pref_t)); // our internal data structure
+	if (prefs == NULL) {
+		fprintf(stderr, "fatal error: failed to calloc prefs structure\n");
+		exit(1);
 	}
-	int num_copied = 0;
-	root = config_root_setting(preferences);
-  setting = config_setting_get_member(root, pref);
-  if(setting && config_setting_is_array(setting)){
-  	int i = 0;
-  	int num_eles = config_setting_length(setting);
-  	num_copied = length;
-  	if(num_eles < length){
-  		num_copied = num_eles;
-  	}
-  	for(i = 0; i < num_copied; ++i){
-  		fillme[i] = config_setting_get_int_elem(setting, i);
-  	}
-  }
-  return num_copied;
-}
 
-const char* preferences_get_metamode_keystrokes(char keystroke, char* pref_key){
-	config_setting_t *keystrokes_group, *root;
-	char setting_str[2] = {keystroke, 0};
-	const char* setting_value;
-	root = config_root_setting(preferences);
-	keystrokes_group = config_setting_get_member(root, pref_key);
-	if(keystrokes_group){
-		if(CONFIG_TRUE == config_setting_lookup_string(keystrokes_group, setting_str, &setting_value)){
-			return setting_value;
+	int is_first_run = 0;
+	
+	config_t config_data; // what libconfig parses out of the file
+	config_t *config = &config_data;
+	config_init(config);
+	
+	if (access(filename, F_OK) == -1) {
+		PRINT(stderr, "Preferences file not found, assuming first run\n");
+		is_first_run = 1;
+	} else {
+		if(config_read_file(config, filename) != CONFIG_TRUE){
+			fprintf(stderr, "%s:%d - %s\n", config_error_file(config),
+			        config_error_line(config), config_error_text(config));
+			char *crash = NULL;
```

Comment:
If the config file has an error, crashing is non-optimal. :-) If a user accidentally hoses their config file (which happens surprisingly often), then they'll be locked out of Term48 and unable to fix it because the app will not start. If the config is broken, we should try to recover, or at least run with a default config and save the broken config file somewhere else so they can try to fix it.